app/lib/frontend/templates/views/shared/search_banner.dart

```diff
@@ -64,3 +76,108 @@ d.Node searchBannerNode({
     ],
   );
 }
+
+/// Create completion data for search input.
+///
+/// Format is dictacted by `pkg/web_app/lib/src/completion.dart`.
+///
+/// If values in `match` is a prefix of what is being typed completion
+/// will automatically start. It'll always try to use the longest match.
+/// If not specified options available are assumed to be `terminal`, that is
+/// they will be followed by whitespace.
+/// If `forcedOnly` is specified, completion can only be initiated with
+/// Ctrl+Space.
+///
+/// This can generate completion data of different sizes given [topics] and
+/// [licenses].
+String completionDataJson({
+  List<String> topics = const [],
+  List<String> licenses = const [],
+}) =>
+    json.encode({
+      'completions': [
```

Comment:
I think these structures should be typed (similarly to what we have in `pkg/_pub_shared/lib/data/`). It can be a follow-up though, let's add a TODO here.

Reply:
yeah, maybe... I kind of wish we didn't have multiple packages...

it'd be much cleaner if we had:
```
lib/src/
  widget/
    completion/
      src/
        ... private bits ...
      widget.dart
      html_as_code.dart
      completions.dart
    widgets.dart  // automatically generated to import widget.dart from all widgets
  ...
```

so that things that are related are located next to each other. Rather than being split out through the code base because they are used in different locations.

But that's a thing I can dream about some other day :rofl: